File: BACKEND/DEVELOP/PYTHON_TEMPLATES/ck_rel.py




################## WARNING - UNTESTED CODE ######################

import sys
import os
sys.path.append(os.getenv('TG_DIR')+"/BACKEND/SCRIPT/LIB_PYTHON/")
import time
from datetime import datetime
import smtplib
import psycopg2
import pickle
import sys
from fct_utils import *
from loader import *

# Get job name
job_name = 'ck_${DATASOURCE}_rel'
JOB_ID = get_job_id_by_name(job_name,True)
JOB_INFO=GLB_TREE[JOB_ID]
PROCESS_CONTROL['JOB_NAME'] =job_name


W_DIR = TG_DIR + GLB_VAR['PROCESS_DIR']
if not os.path.isdir(W_DIR):
	fail_process(JOB_ID + '001', 'NO ' + W_DIR + ' found ')

W_DIR +=  JOB_INFO['DIR'] 
if not os.path.isdir(W_DIR):
	os.mkdir(W_DIR)
	if not os.path.isdir(W_DIR):
		fail_process(JOB_ID + '002', 'Unable to find and create ' + W_DIR)

os.chdir(W_DIR)

if (os.getcwd() != W_DIR):	
	fail_process(JOB_ID + '003', 'Unable to chdir ' + W_DIR+"\n Current directory: "+os.getcwd())

add_log("Working directory:"+W_DIR)


add_log("Get FTP_${DATASOURCE} link and ${DATASOURCE} release date")
# Get FTP_${DATASOURCE} link
if not 'FTP_${DATASOURCE}' in GLB_VAR['LINK']:
	fail_process(JOB_ID + '004', 'FTP_${DATASOURCE} not found')

# Download index file from FTP_${DATASOURCE}
add_log("Download version from FTP_${DATASOURCE}")



# Modify path to download file containing version of the data source
#if (not dl_file(GLB_VAR['LINK']['FTP_${DATASOURCE}']+'/Flat_file_tab_delimited/' ,3, 'index.html')):
#	fail_process(JOB_ID + '005', 'Unable to download index.html from FTP_${DATASOURCE}')

# Get the last update date from index.html
add_log("Get the last update date from index.html")

new_release_date=''

#ADD PROCESS  TO GET THE LATEST VERSION DATE

if (new_release_date == ''):
	fail_process(JOB_ID + '007', 'Unable to extract date from index.html')

add_log("New release date:"+new_release_date)




# Compare the last update date with the one in the database
add_log("Compare the last update date with the one in the database")
CURR_RELEASE=get_current_release_date('${DATASOURCE_NC}',JOB_ID)


if (CURR_RELEASE == new_release_date):
	add_log("Same as current release date")
	# No new release found
	# Update
	success_process("VALID")


add_log("Update release tag due to new release date")
update_release_date(JOB_ID,'${DATASOURCE_NC}',new_release_date)


add_log("Create working directory")
PROCESS_CONTROL['DIR']='N/A'

today_date=get_curr_date()
# Create directory
PROCESS_CONTROL['DIR'] = today_date
if (not os.path.isdir(today_date)):
	os.mkdir(today_date)
	if not os.path.isdir(today_date):
		fail_process(JOB_ID + '009', 'Unable to create ' + today_date + ' directory')

success_process()

File: BACKEND/DEVELOP/PHP_TEMPLATES/ck_rel.php
<?php

/**
 SCRIPT NAME: ck_${DATASOURCE}_rel
 PURPOSE:     Check for new release of the ${DATASOURCE}
 
*/
$JOB_NAME='ck_${DATASOURCE}_rel';

/// Get root directories
$TG_DIR= getenv('TG_DIR');
if ($TG_DIR===false)  die('NO TG_DIR found ');
if (!is_dir($TG_DIR)) die('TG_DIR value is not a directory '.$TG_DIR);
require_once($TG_DIR.'/BACKEND/SCRIPT/LIB/loader.php');

/// Get job id
$JOB_ID=getJobIDByName($JOB_NAME);
$PROCESS_CONTROL['DIR']='N/A';
/// Get job info
$JOB_INFO=$GLB_TREE[$JOB_ID];


addLog("Check variables");
	/// Setting up working directory:
	$W_DIR=$TG_DIR.'/'.$GLB_VAR['PROCESS_DIR'];if (!is_dir($W_DIR)) 					failProcess($JOB_ID."001",'NO '.$W_DIR.' found ');
	$W_DIR.='/'.$JOB_INFO['DIR'].'/';	   	   if (!is_dir($W_DIR) && !mkdir($W_DIR)) 	failProcess($JOB_ID."002",'Unable to find and create '.$W_DIR);
	if (!chdir($W_DIR))																	failProcess($JOB_ID."003",'Unable to chdir '.$W_DIR);
	
	/// Checking ftp file paths:
	if (!isset($GLB_VAR['LINK']['FTP_${DATASOURCE}']))										failProcess($JOB_ID."004",'FTP_${DATASOURCE} path no set');

addLog("Check license from ${DATASOURCE}");	

	/// Download the license file for this data source
	/// and compare it to the current version of the license stored

	if (is_file($W_DIR.'/LICENSE') && !unlink($W_DIR.'/LICENSE'))						failProcess($JOB_ID."009",'Unable to delete LICENSE');
	
	/// Please adapt here the path to the license:
	if (!dl_file($GLB_VAR['LINK']['FTP_${DATASOURCE}'].'/LICENSE',3,'LICENSE'))		failProcess($JOB_ID."010",'Unable to download license file');
	
	/// First time: set it as the license
	if (!is_file('CURRENT_LICENSE'))
	{
		rename('LICENSE','CURRENT_LICENSE');
	}
	else 
	{
		addLog("Compare License");
		/// Otherwise we compare line by line
		$fp=fopen('LICENSE','r'); if (!$fp)												failProcess($JOB_ID."010",'Unable to open LICENSE');
		$fp2=fopen('CURRENT_LICENSE','r'); if (!$fp2)									failProcess($JOB_ID."011",'Unable to open CURRENT_LICENSE');
		$VALID=true;
		while(!feof($fp))
		{
			$line=stream_get_line($fp,10000,"\n");
			$line2=stream_get_line($fp2,10000,"\n");
			if ($line!=$line2){$VALID=false;break;}
		}
		fclose($fp);
		fclose($fp2);
		if (!unlink($W_DIR.'/LICENSE'))													failProcess($JOB_ID."012",'Unable to remove LICENSE');
		if (!$VALID)																	failProcess($JOB_ID."013",'License file is different');
	}	


addLog("Get release date from ${DATASOURCE}");
	
	/// Please adapt here the path to the license:
	if (!dl_file($GLB_VAR['LINK']['FTP_${DATASOURCE}'].'/VERSION',3,'VERSION'))		failProcess($JOB_ID."010",'Unable to download VERSION file');

	/// Process file to retrieve the license
	
	
addLog("Get current release date for ${DATASOURCE}");
	$CURR_RELEASE=getCurrentReleaseDate('${DATASOURCE_NC}',$JOB_ID);
	
addLog("Compare release date ".$CURR_RELEASE."\t".$NEW_RELEASE);
	if ($CURR_RELEASE == $NEW_RELEASE)
	{
		successProcess('VALID');
	}
	

addLog("Update release tag for ${DATASOURCE}");
	updateReleaseDate($JOB_ID,'${DATASOURCE_NC}',$NEW_RELEASE);

	
addLog("Create directory");
	$PROCESS_CONTROL['DIR']='N/A';

	/// Create directory:
	$W_DIR=$TG_DIR.'/'.$GLB_VAR['PROCESS_DIR'];if (!is_dir($W_DIR)) 					failProcess($JOB_ID."010",'NO '.$W_DIR.' found ');
	$W_DIR.='/'.$JOB_INFO['DIR'].'/';	   if (!is_dir($W_DIR) && !mkdir($W_DIR)) 		failProcess($JOB_ID."011",'Unable to find and create '.$W_DIR);
	$W_DIR.=getCurrDate();		           if (!is_dir($W_DIR) && !mkdir($W_DIR)) 		failProcess($JOB_ID."012",'Unable to create new process dir '.$W_DIR);

	/// Update process control directory to the current release so that the next job can use it
	$PROCESS_CONTROL['DIR']=getCurrDate();
	


successProcess();

?>
